nomad/eval_endpoint.go

```diff
@@ -363,45 +363,34 @@ func (e *Eval) List(args *structs.EvalListRequest,
 			}
 
 			var evals []*structs.Evaluation
-			var itemCount int32
-			var lastToken string
-			foundToken := args.QueryOptions.LastToken == ""
+			pageTracker := newListPageTracker(args.QueryOptions)
+
 			for {
 				raw := iter.Next()
 				if raw == nil {
-					lastToken = ""
 					break
```

Comment:
This is great! One non-blocking question though. Are we going to include some sort of `EOF` indicator?  I think it would have get set external to the tracker, and this seems like the place.  Unless, everything moved inside the tracker and it could handle this itself.

Reply:
Unfortunately no. The API doesn't wrap the list results with a `api.QueryMeta`, so there's no token actually exposed to the caller. We can only rely on the caller getting the _last_ token by checking the last token it got in the `[]*api.Evaluation`. So the API client always has to grab the next (empty) page to find out if it's run out of results. That stinks but it's the only way to do it without changing the shape of the API object.

If we had a "next" token, that's where we'd want to return some kind of EOF value, and that's what I'd want to do in some hypothetical API v2.

Reply:
_Hold on a second!_ We set a `X-Nomad-LastToken` header in the response. Which means we don't need to change the shape of the API at all and we can return the `NextToken` instead and get proper empty set checking. Boo, I wish I'd realized that before changing all this. But easy enough to fix.

Reply:
Yeah. I think you could probably put everything inside the Tracker too and let it handle things.  Add a `resolveTokenFn` to the args similar to `MatchesFilter`, then pass the args to the tracker, then I think you'd have Inversion of Control, and it could give you back a page/[]interface{} and all the boilerplate would be inside the tracker and the List methods wouldn't have for loops.

Reply:
I'm going to rework the paginator so it's like this https://play.golang.com/p/_RRJy0y-OFa

Reply:
I'm going to implement the paginator something like: https://play.golang.com/p/QJPlxM_m0di